api/rest/coordinator.go

```diff
@@ -0,0 +1,231 @@
+	mergedZip, err := ioutil.TempFile("", bundleName)
+	if err != nil {
+		return "", err
+	}
+	defer mergedZip.Close()
+
+	zipWriter := zip.NewWriter(mergedZip)
+	defer zipWriter.Close()
+
+	for _, p := range bundlePaths {
+		err = appendToZip(zipWriter, p)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return mergedZip.Name(), nil
+}
+
+func appendToZip(writer *zip.Writer, path string) error {
+	r, err := zip.OpenReader(path)
+	if err != nil {
+		return fmt.Errorf("could not open %s: %s", path, err)
+	}
+	defer r.Close()
+
+	for _, f := range r.File {
+		rc, err := f.Open()
+		if err != nil {
+			return fmt.Errorf("could not open %s from zip: %s", f.Name, err)
+		}
+
+		file, err := writer.Create(f.Name)
+		if err != nil {
+			return fmt.Errorf("could not create file %s: %s", f.Name, err)
+		}
+		_, err = io.Copy(file, rc)
+		if err != nil {
+			return fmt.Errorf("could not copy file %s to zip: %s", f.Name, err)
+		}
+		rc.Close()
+	}
+
+	return nil
+}
+
+func (c ParallelCoordinator) createBundle(ctx context.Context, node node, id string, jobs chan<- job) BundleStatus {
+	_, err := c.client.CreateBundle(ctx, node.baseURL, id)
+	if err != nil {
+		// Return done status with error. To mark node as errored so file will not be downloaded
+		return BundleStatus{
+			ID:   id,
+			node: node,
+			done: true,
+			err:  fmt.Errorf("could not create bundle: %s", err),
+		}
+	}
+
+	// Schedule bundle status check
+	jobs <- func() BundleStatus {
+		return c.waitForDone(ctx, node, id, jobs)
+	}
+
+	// Return undone status with no error.
+	return BundleStatus{ID: id, node: node}
+}
+
+func (c ParallelCoordinator) waitForDone(ctx context.Context, node node, id string, jobs chan<- job) BundleStatus {
+	statusCheck := func() {
+		jobs <- func() BundleStatus {
+			return c.waitForDone(ctx, node, id, jobs)
+		}
+	}
+
+	//TODO(janisz): Handle context
+
+	logrus.WithField("IP", node.IP).Info("Checking bundle status on node.")
+	// Check bundle status
+	bundle, err := c.client.Status(ctx, node.baseURL, id)
+	// If error
+	if err != nil {
+		// then schedule next check in given time.
+		// It will only add check to job queue so interval might increase but it's OK.
+		time.AfterFunc(time.Second, statusCheck)
+		// Return status with error. Do not mark bundle as done yet. It might change it status
+		return BundleStatus{ID: id, node: node, err: fmt.Errorf("could not check status: %s", err)}
+	}
+	// If bundle is in terminal state (its state won't change)
+	if bundle.Status == Done || bundle.Status == Deleted || bundle.Status == Canceled {
+		logrus.WithField("IP", node.IP).Info("Node bundle is finished.")
+		// mark it as done
+		return BundleStatus{ID: id, node: node, done: true}
+	}
+	// If bundle is still in progress (InProgress, Unknown or Started)
+	// then schedule next check in given time
+	// It will only add check to job queue so interval might increase but it's OK.
+	time.AfterFunc(time.Second, statusCheck)
```

Comment:
Make it configurable. At production it should be couple of seconds but in tests it should be millisecond.